Plugin/module docs: parse return values, add collection names in them (version_added_collection), and format them nicely in ansible-doc

felixfontein opened the pull request:
##### SUMMARY
Right now, return value docs are treated as a string until the final stages (ansible-doc simply dumps them to the output, and plugin_formatter tries to parse the YAML). In particular, `version_added` in there are not tagged or untagged. (Which isn't really needed, since the versions should always refer to the current collection / ansible-base.)

This PR modifies the plugin docs processing so that the return value docs are also parsed in lib/ansible/parsing/plugin_docs.py and also tagged/untagged. Finally, it modifies ansible-doc so that it formats them nicer (using the same mechanism as for plugin/module options).

Combined with #69795 this gives a lot nicer representation of return values; for example (`ansible-doc acme_certificate_info`):
```
- account_uri
        ACME account URI, or None if account does not exist.

        returned: always
        type: str

- exists
        Whether the account exists.

        returned: always
        type: bool
```
instead of
```
exists:
  description: Whether the account exists.
  returned: always
  type: bool

account_uri:
  description: ACME account URI, or None if account does not exist.
  returned: always
  type: str
```
Also note the ordering - instead of just dumping the return values as described in the return value string, they will be sorted.

##### ISSUE TYPE
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
ansible-doc

felixfontein committed:
Tag return value docs if they are a dict (and not str/None).

felixfontein committed:
Try to parse return docs as YAML.

felixfontein committed:
Properly dump return values in ansible-doc.

felixfontein committed:
Adjust plugin formatter.

felixfontein committed:
Add changelog fragment.

felixfontein committed:
Don't add 'default' for return values.

felixfontein committed:
Fix plugin_formatter.

felixfontein committed:
Only try to parse return docs if they are still a string.

felixfontein commented:
@samdoran for modules, validate-modules does validate it to some regard (validation was improved for 2.9 and 2.10, and a lot of ignore.txt entries added IIRC). So it shouldn't be too bad.

Will add tests.

felixfontein committed:
Add tests.

samdoran commented:
@felixfontein I mainly just wanted to make sure this code is bit for flexible with the data it gets from the RETURN section so it doesn't explode when/if it encounters something odd.

felixfontein committed:
Warn if RETURN cannot be parsed.

felixfontein commented:
Discussion from meeting (TBC): warn in case of unparsable YAML, or fail?

Note that this function is used in only three situations:
1. by ansible-doc
2. by ansible-console's help (it only cares for DOCUMENTATION and ignores RETURN)
3. by plugin_formatter (which will vanish soon)

bcoca commented on lib/ansible/parsing/plugin_docs.py:
```diff
@@ -57,6 +57,15 @@ def read_docstring(filename, verbose=True, ignore_errors=True):
                             if theid == 'DOCUMENTATION':
                                 # string should be yaml
                                 data[varkey] = AnsibleLoader(child.value.s, file_name=filename).get_single_data()
+                            elif theid == 'RETURN':
```
i would jsut put both in prev conditional `if theid in ('DOCUMENTATION, 'RETURN')`

felixfontein replied:
As long as there is no consensus that bad YAML should result in an error, I'll keep it this way.

bcoca commented on test/integration/targets/ansible-doc/library/test_docs_returns.py:
```diff
@@ -0,0 +1,60 @@
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
remove

felixfontein replied:
done

felixfontein committed:
Adjust tests. Also test for warning.

mattclay commented on lib/ansible/parsing/plugin_docs.py:
```diff
@@ -56,10 +56,19 @@ def read_docstring(filename, verbose=True, ignore_errors=True):
-                            if theid in ['EXAMPLES', 'RETURN']:
-                                # examples 'can' be yaml, return must be, but even if so, we dont want to parse as such here
+                            if theid == 'EXAMPLES':
+                                # examples 'can' be yaml, but even if so, we dont want to parse as such here
```
In ansible-base we've required examples to be valid yaml for quite some time:

https://github.com/ansible/ansible/blob/840d3a9dd7d1233ab55aa9516fc9b44e4c865a81/test/lib/ansible_test/_data/sanity/yamllint/yamllinter.py#L171-L175

felixfontein replied:
I guess the main concern are third-party plugins that weren't tested with `ansible-test`. OTOH this only affects `ansible-doc` and `ansible-console`'s help, so I think it is an acceptable risk.

bcoca replied:
@mattclay  also, this is not just modules, but other plugins, you sometimes have an 'ini file' or other 'non yaml content' that goes in this section, so we should not restrict at docs code.

felixfontein replied:
@bcoca I *think* it checks all plugins, not just modules (despite the naming). Would have to check though...

mattclay replied:
@felixfontein Yes, the `yamllint` sanity test does check the docs on all plugin types except `module_utils`, despite the name of the function being `check_modules`.

bcoca replied:
@mattclay have we made exceptions for certain plugins? some are using ini files in examples ?

mattclay replied:
Not that I'm aware of.

bcoca replied:
```
EXAMPLES = r'''
    # define 2 hosts in command line
    # ansible -i '10.10.2.6, 10.10.2.4' -m ping all

    # DNS resolvable names
    # ansible -i 'host1.example.com, host2' -m user -a 'name=me state=absent' all

    # just use localhost
    # ansible-playbook -i 'localhost,' play.yml -c local
```
seems people have been 'creative' to avoid the restriction, we really should not make this a strict requirement

bcoca replied:
specially since nothing will display if we treat it as yaml (currently its processed as string)

bcoca replied:
```
EXAMPLES = '''
  example1: |
      # example cfg file
      [web]
      host1
      host2 ansible_port=222 # defined inline, interpreted as an integer

      [web:vars]
```

abadger replied:
I know that examples is not valid structured yaml for a large number of plugins.  Perhaps yamllint decides it's a valid yaml string?

felixfontein committed:
if -> elif (otherwise EXAMPLE will be parsed too).

felixfontein committed:
Always parse return documentation, and fail if it is invalid YAML.

felixfontein commented:
I've adjusted the PR to always parse return docs as YAML, and fail if they cannot be parsed. A module with invalid return docs will return the following error in ansible-doc:
```
$ ansible-doc --playbook-dir ./ test_docs_returns_broken
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying
out features under development. This is a rapidly changing source of code and can become unstable at any point.
ERROR! module test_docs_returns_broken missing documentation (or could not parse documentation): while parsing a block mapping
  in "<unicode string>", line 2, column 1
did not find expected key
  in "<unicode string>", line 4, column 4
```
Since all modules passing `ansible-test sanity` have valid return docs (see @mattclay's comment), I think it is acceptable to require this also for third-party modules. Especially since it only affects `ansible-doc` and `ansible-console`'s `help` command.

acozine commented:
Discussed at the Docs Working Group meeting today, unanimously approved.

felixfontein committed:
Polishing.

felixfontein commented:
I've rebased the PR after #69926 was merged. This simplifies several parts, and also addresses the untagging placement in ansible-doc that @bcoca mentioned on IRC. (It's now in a better place, and also right now commented out - whether to re-enable it, and what exactly it should do, can be debated later, since it is not related to this PR.)

ready_for_review

felixfontein committed:
Mostly re-enable ansible-doc tests.

Listing from the local collection seems to be somewhat broken. I assume this
is why the test was disabled.

bcoca commented on test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py:
```diff
@@ -6,7 +6,7 @@
 
 DOCUMENTATION = '''
-    cache: testns.testcol.notjsonfile
+    cache: notjsonfile
```
this will create issues when loading the plugin, they should keep  FQCN name in docs

felixfontein replied:
Hmm, then why are there no FQCNs in community.general for plugins?

felixfontein replied:
I've checked, and the plugin loader ignores it and doesn't pass this part of DOCUMENTATION on. The only place which seems to be using this is ansible-doc's man formatter, which uses it for the title.

Since it does the same for modules, and there FQCN is explicitly forbidden by ansible-test sanity, I think having the short name here is the better choice.

felixfontein replied:
@bcoca if you insist, I can still change it. Just tell me.

mattclay replied:
If the FQCN fails the 2.9 sanity tests then requiring it for 2.10 will create an issue for collections trying to support both versions.

felixfontein replied:
For modules, FQCN will fail both the 2.9 and 2.10 tests (if these aren't changed). For plugins, there is no test.

@abadger mentioned in #ansible-devel that the docs pipeline kind of depends (right now) on shortnames being used.

felixfontein committed:
Lint and make tests work with Python 2.

s-hertel commented on test/integration/targets/ansible-doc/runme.sh:
```diff
@@ -12,18 +12,13 @@ current_out="$(ansible-doc --playbook-dir ./ testns.testcol.fakemodule)"
 test "$current_out" == "$expected_out"
 
-# test module docs from collection
-current_out="$(ansible-doc --playbook-dir ./ test_docs_suboptions)"
```
I like having the additional --playbook-dir coverage since I think this was catching a legitimate bug.

oh, I overlooked the one right above it - nevermind me

felixfontein replied:
This one actually didn't work because of unicode characters in Python 2. That's why I moved it (and the ones below, which I added in this PR) into the playbook.

s-hertel replied:
Ah, okay. That does seem like something we should fix (though I wouldn't hold this one up for it).

felixfontein replied:
Yes, and also the listing counting tests I had to comment out. :)

s-hertel reviewed (commented):
Looks good to me. It looks like the collection finder tries to load files that don't exist and requires absolute paths, so with this diff the collection plugin type tests pass again https://gist.github.com/s-hertel/1dc8415e3f5b24cd62c456eba1c0f83b.

felixfontein committed:
Keep FQCNs in plugins (not modules), i.e. restore previous state.

felixfontein commented on test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/vars/noop_vars_plugin.py:
```diff
@@ -2,7 +2,7 @@
 
 DOCUMENTATION = '''
-    vars: noop_vars_plugin
+    vars: testns.testcol.noop_vars_plugin
```
@bcoca this wasn't a FQCN before apparently.

felixfontein commented:
/rebuild_partial

felixfontein commented:
@samdoran @acozine @abadger @mattclay @bcoca @s-hertel thanks for reviewing and merging!
